pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/function/BaseTransformFunction.java

```diff
@@ -0,0 +1,511 @@
+        for (int i = 0; i < length; i++) {
+          int numValues = intValues[i].length;
+          _doubleValuesMV[i] = new double[numValues];
+          copyToDouble(intValues[i], _doubleValuesMV[i], numValues);
+        }
+        return _doubleValuesMV;
+      case LONG:
+        long[][] longValues = transformToLongValuesMV(projectionBlock);
+        for (int i = 0; i < length; i++) {
+          int numValues = longValues[i].length;
+          _doubleValuesMV[i] = new double[numValues];
+          copyToDouble(longValues[i], _doubleValuesMV[i], numValues);
+        }
+        return _doubleValuesMV;
+      case FLOAT:
+        float[][] floatValues = transformToFloatValuesMV(projectionBlock);
+        for (int i = 0; i < length; i++) {
+          int numValues = floatValues[i].length;
+          _doubleValuesMV[i] = new double[numValues];
+          copyToDouble(floatValues[i], _doubleValuesMV[i], numValues);
+        }
+        return _doubleValuesMV;
+      default:
+        throw new UnsupportedOperationException();
+    }
+  }
+
+  @Override
+  public String[][] transformToStringValuesMV(@Nonnull ProjectionBlock projectionBlock) {
+    if (_stringValuesMV == null) {
+      _stringValuesMV = new String[DocIdSetPlanNode.MAX_DOC_PER_CALL][];
+    }
+    int length = projectionBlock.getNumDocs();
+    switch (getResultMetadata().getDataType()) {
+      case INT:
+        int[][] intValues = transformToIntValuesMV(projectionBlock);
+        for (int i = 0; i < length; i++) {
+          int numValues = intValues[i].length;
+          _stringValuesMV[i] = new String[numValues];
+          copyToString(intValues[i], _stringValuesMV[i], numValues);
+        }
+        return _stringValuesMV;
+      case LONG:
+        long[][] longValues = transformToLongValuesMV(projectionBlock);
+        for (int i = 0; i < length; i++) {
+          int numValues = longValues[i].length;
+          _stringValuesMV[i] = new String[numValues];
+          copyToString(longValues[i], _stringValuesMV[i], numValues);
+        }
+        return _stringValuesMV;
+      case FLOAT:
+        float[][] floatValues = transformToFloatValuesMV(projectionBlock);
+        for (int i = 0; i < length; i++) {
+          int numValues = floatValues[i].length;
+          _stringValuesMV[i] = new String[numValues];
+          copyToString(floatValues[i], _stringValuesMV[i], numValues);
+        }
+        return _stringValuesMV;
+      case DOUBLE:
+        double[][] doubleValues = transformToDoubleValuesMV(projectionBlock);
+        for (int i = 0; i < length; i++) {
+          int numValues = doubleValues[i].length;
+          _stringValuesMV[i] = new String[numValues];
+          copyToString(doubleValues[i], _stringValuesMV[i], numValues);
+        }
+        return _stringValuesMV;
+      default:
+        throw new UnsupportedOperationException();
+    }
+  }
+
+  private static void copyToInt(long[] source, int[] dest, int length) {
```

Comment:
move all this methods to some util class. Also add comments on why we cant use Arrays.copy methods

Reply:
Done